packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -145,66 +168,407 @@ String getGradleVersionForAndroidPlugin(Directory directory, Logger logger) {
+  }
+  // System installed Gradle version.
+  if (processManager.canRun('gradle')) {
+    final String gradleVersionVerbose =
+        (await processManager.run(<String>['gradle', gradleVersionFlag])).stdout
+            as String;
+    // Expected format:
+/*
+
+------------------------------------------------------------
+Gradle 7.6
+------------------------------------------------------------
+
+Build time:   2022-11-25 13:35:10 UTC
+Revision:     daece9dbc5b79370cc8e4fd6fe4b2cd400e150a8
+
+Kotlin:       1.7.10
+Groovy:       3.0.13
+Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
+JVM:          17.0.6 (Homebrew 17.0.6+0)
+OS:           Mac OS X 13.2.1 aarch64
+    */
+    // Observation shows that the version can have 2 or 3 numbers.
+    // Inner parentheticals `(\.\d+)?` denote the optional third value.
+    // Outter parentheticals `Gradle (...)` denote a grouping used to extract
+    // the version number.
+    final RegExp gradleVersionRegex = RegExp(r'Gradle (\d+\.\d+(\.\d+)?)');
+    final RegExpMatch? version =
+        gradleVersionRegex.firstMatch(gradleVersionVerbose);
+    if (version == null) {
+      // Most likley a bug in our parse implementation/regex.
+      logger.printWarning(_formatParseWarning(gradleVersionVerbose));
+      return null;
+    }
+    return version.group(1);
+  } else {
+    logger.printTrace('Could not run system gradle');
+    return null;
+  }
 }
 
+/// Returns the Android Gradle Plugin (AGP) version that the current project
+/// depends on when found, null otherwise.
+///
+/// The Android plugin version is specified in the [build.gradle] file within
+/// the project's Android directory ([androidDirectory]).
+String? getAgpVersion(Directory androidDirectory, Logger logger) {
+  final File buildFile = androidDirectory.childFile('build.gradle');
+  if (!buildFile.existsSync()) {
+    logger.printTrace('Can not find build.gradle in $androidDirectory');
+    return null;
+  }
+  final String buildFileContent = buildFile.readAsStringSync();
+  final Iterable<Match> pluginMatches =
+      _androidGradlePluginRegExp.allMatches(buildFileContent);
+  if (pluginMatches.isEmpty) {
+    logger.printTrace("$buildFile doesn't provide an AGP version");
+    return null;
+  }
+  final String? androidPluginVersion = pluginMatches.first.group(1);
+  logger.printTrace('$buildFile provides AGP version: $androidPluginVersion');
+  return androidPluginVersion;
+}
+
+String _formatParseWarning(String content) {
+  return 'Could not parse gradle version from: \n'
+      '$content \n'
+      'If there is a version please look for an existing bug '
+      'https://github.com/flutter/flutter/issues/'
+      ' and if one does not exist file a new issue.';
+}
+
+// Validate that Gradle version and AGP are compatible with each other.
+//
+// Returns true if versions are compatible.
+// Null Gradle version or AGP version returns false.
+// If compatability can not be evaulated returns false.
```

Comment:
Should we use a tri-state return instead? It seems like "we know these are not compatible" and "we don't know if these are compatible because they are too new for use to know anything about" are different enough to have distinct handling.